tests/unit/utils/test_wsgi.py

```diff
@@ -168,6 +176,39 @@ def test_passes_through_headers(self):
         assert app.calls == [pretend.call({"HTTP_X_FOOBAR": "wat"}, start_response)]
 
 
+def test_ip_address_exists(db_request):
+    ip_address = DBIpAddressFactory(ip_address="192.0.2.69")
+    db_request.environ["REMOTE_ADDR"] = "192.0.2.69"
+    db_request.remote_addr = "192.0.2.69"
+
+    assert wsgi._ip_address(db_request) == ip_address
+
+
+def test_ip_address_created(db_request):
+    with pytest.raises(NoResultFound):
+        db_request.db.query(IpAddress).filter(
+            IpAddress.ip_address == "192.0.2.69"
+        ).one()
+
+    db_request.environ["REMOTE_ADDR"] = "192.0.2.69"
+    db_request.environ["GEOIP_CITY"] = "Anytown, ST"
+    db_request.remote_addr = "192.0.2.69"
+
+    wsgi._ip_address(db_request)
+
+    ip_address = (
+        db_request.db.query(IpAddress)
+        .filter(IpAddress.ip_address == "192.0.2.69")
+        .one()
+    )
+    assert ip_address.ip_address == "192.0.2.69"
+    assert (
+        ip_address.hashed_ip_address
+        == "6694f83c9f476da31f5df6bcc520034e7e57d421d247b9d34f49edbfc84a764c"
```

Comment:
```suggestion
        ip_address.hashed_ip_address == remote_addr_hashed
```
This has value is of `remote_addr` since that's being put into the test request over here: https://github.com/pypi/warehouse/blob/352fec979299c1d64551ca4b2c1622e896e2adb7/tests/conftest.py#L180

So there's something amiss about this specific test, since it's comparing a hash that isn't for the locally set IP address.

Likely needs an entry after `db_request.environ["REMOTE_ADDR"] =` to set the hashed value.

Alternately, replace `db_request` with `db_session` fixture, which has less request scaffolding applied if you can?

Reply:
https://github.com/pypi/warehouse/pull/13594/commits/d9357aaf68aed31877e9c717849dc72334599860